File: PasswordGenerator/PHP/Cli/password_generator.php
<?php

function generate_password(int $length = 16, bool $use_upper = true, bool $use_lower = true, bool $use_digits = true, bool $use_symbols = true): string {
    if ($length < 1) {
        throw new InvalidArgumentException("Password length must be at least 1");
    }

    $upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    $lower = "abcdefghijklmnopqrstuvwxyz";
    $digits = "0123456789";
    $symbols = "!@#$%^&*()_+-=[]{}|;:,.<>?";

    $categories = [];
    if ($use_upper) $categories[] = $upper;
    if ($use_lower) $categories[] = $lower;
    if ($use_digits) $categories[] = $digits;
    if ($use_symbols) $categories[] = $symbols;

    if (empty($categories)) {
        throw new InvalidArgumentException("At least one character category must be enabled");
    }

    if ($length < count($categories)) {
        throw new InvalidArgumentException(
            "Password length must be at least " . count($categories) .
            " when " . count($categories) . " categories are enabled"
        );
    }

    $password = "";
    foreach ($categories as $cat) {
        $password .= $cat[random_int(0, strlen($cat) - 1)];
    }

    $all_chars = implode("", $categories);
    $remaining = $length - count($categories);
    for ($i = 0; $i < $remaining; $i++) {
        $password .= $all_chars[random_int(0, strlen($all_chars) - 1)];
    }

    $password = str_split($password);
    shuffle($password);
    return implode("", $password);
}

function show_menu(): void {
    echo "=== Password Generator ===\n";
    echo "Length (default 16): ";
    $input = trim(fgets(STDIN));
    $length = $input === "" ? 16 : (int)$input;

    echo "Include uppercase? (Y/n): ";
    $use_upper = strtolower(trim(fgets(STDIN))) !== "n";

    echo "Include lowercase? (Y/n): ";
    $use_lower = strtolower(trim(fgets(STDIN))) !== "n";

    echo "Include digits? (Y/n): ";
    $use_digits = strtolower(trim(fgets(STDIN))) !== "n";

    echo "Include symbols? (Y/n): ";
    $use_symbols = strtolower(trim(fgets(STDIN))) !== "n";

    try {
        $password = generate_password($length, $use_upper, $use_lower, $use_digits, $use_symbols);
        echo "\nGenerated password: $password\n";
    } catch (InvalidArgumentException $e) {
        echo "Error: " . $e->getMessage() . "\n";
    }
}

if (PHP_SAPI === "cli" && isset($argv[0]) && realpath($argv[0]) === __FILE__) {
    if ($argc > 1) {
        $opts = getopt("l:", ["no-upper", "no-lower", "no-digits", "no-symbols", "length:"]);
        $length = isset($opts["l"]) ? (int)$opts["l"] : (isset($opts["length"]) ? (int)$opts["length"] : 16);
        $use_upper = !isset($opts["no-upper"]);
        $use_lower = !isset($opts["no-lower"]);
        $use_digits = !isset($opts["no-digits"]);
        $use_symbols = !isset($opts["no-symbols"]);
        try {
            echo generate_password($length, $use_upper, $use_lower, $use_digits, $use_symbols) . "\n";
        } catch (InvalidArgumentException $e) {
            fwrite(STDERR, "Error: " . $e->getMessage() . "\n");
            exit(1);
        }
    } else {
        show_menu();
    }
}

File: PasswordGenerator/CSharp/Cli/PasswordGenerator.cs
namespace PasswordGeneratorCli;

public static class PasswordGenerator
{
    private const string Upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private const string Lower = "abcdefghijklmnopqrstuvwxyz";
    private const string Digits = "0123456789";
    private const string Symbols = "!@#$%^&*()_+-=[]{}|;:,.<>?";

    public static string Generate(int length = 16, bool useUpper = true, bool useLower = true,
        bool useDigits = true, bool useSymbols = true)
    {
        if (length < 1)
            throw new ArgumentException("Password length must be at least 1");

        var categories = new List<string>();
        if (useUpper) categories.Add(Upper);
        if (useLower) categories.Add(Lower);
        if (useDigits) categories.Add(Digits);
        if (useSymbols) categories.Add(Symbols);

        if (categories.Count == 0)
            throw new ArgumentException("At least one character category must be enabled");

        if (length < categories.Count)
            throw new ArgumentException(
                $"Password length must be at least {categories.Count} " +
                $"when {categories.Count} categories are enabled");

        var rng = Random.Shared;
        var passwordChars = new char[length];
        var index = 0;

        foreach (var cat in categories)
            passwordChars[index++] = cat[rng.Next(cat.Length)];

        var allChars = string.Concat(categories);
        while (index < length)
            passwordChars[index++] = allChars[rng.Next(allChars.Length)];

        rng.Shuffle(passwordChars);
        return new string(passwordChars);
    }
}
